www/src/public/js/controllers/builder/main.js

```diff
@@ -2070,6 +2308,332 @@
+                    $scope.selectedList.items[cslot].id = -5;
+                    break;
+                case 14:
+                    charm3 += this.charm1Selector;
+                    charm3 += this.charm2Selector;
+                    charm3 += this.charm3Selector;
+                    charm3 += this.charm4Selector;
+                    charm3 += this.charm5Selector;
+
+                    $scope.selectedList.runeCharms.charm3 = charm3;
+
+                    var runeStats = applyRunecharmStats($scope.selectedList.runeCharms.charm3);
+                    for (const [stat, num] of Object.entries(runeStats)) {
+                        $scope.selectedList.items[cslot][stat] = num;
+                    }
+                    var name = runeStats.charmName.substring(0, runeStats.charmName.length-1);
+
+                    $scope.selectedList.items[cslot].name = "Runecharm ("+ name + ")";
+                    $scope.selectedList.items[cslot].id = -5;
+                    break;
+                case 15:
+                    charm4 += this.charm1Selector;
+                    charm4 += this.charm2Selector;
+                    charm4 += this.charm3Selector;
+                    charm4 += this.charm4Selector;
+                    charm4 += this.charm5Selector;
+
+                    $scope.selectedList.runeCharms.charm4 = charm4;
+                    var runeStats = applyRunecharmStats($scope.selectedList.runeCharms.charm4);
+                    for (const [stat, num] of Object.entries(runeStats)) {
+                        $scope.selectedList.items[cslot][stat] = num;
+                    }
+                    var name = runeStats.charmName.substring(0, runeStats.charmName.length-1);
+
+                    $scope.selectedList.items[cslot].name = "Runecharm ("+ name + ")";
+                    $scope.selectedList.items[cslot].id = -5;
+                    break;
+            }
+            $scope.saveClientSideData();
+            //applyItemRestrictions();
+        };
+
+        /** applies stats to a runecharm from a given string **/
+        var applyRunecharmStats = function(charmStr) {
+            
+            var runeStats = { 
+                "strength": 0,
+                "mind": 0,
+                "dexterity": 0,
+                "constitution": 0,
+                "perception": 0,
+                "spirit": 0,
+                "hit": 0,
+                "dam": 0,
+                "hp": 0,
+                "ma": 0,
+                "mv": 0,
+                "hpr": 0,
+                "mar": 0,
+                "mvr": 0,
+                "spellcrit": 0,
+                "spelldam": 0,
+                "ac": -5,
+                "rangedAccuracy": 0,
+                "manaReduction": 0,
+                "concentration": 0,
+                "rent": 225,
+                "charmName": ""
+            };
+
+            for(var i = 0; i < charmStr.length; ++i) {
+                switch (charmStr[i]) {
```

Comment:
Rather than have this giant switch/case statement, it might be useful to include the stat change amount and charm name in the `$scope.charmOptions` objects. That way this giant block of code can turn into:

```
for (var i = 0; i < charmStr.length; ++i) {
  if (charmStr[i] == "A") continue;
  var charmOption = $scope.charmOptions[charmStr[i]];
  runeStats.rent += 203;
  runeStats[charmOption.statVar] += charmOption.statValue;
  runeStats.charmName += charmOption.charmName + "/";
}
```

Much cleaner, no?

Reply:
One change that will have to happen to support this, however, is to set up your `$scope.charmOptions` as:
```
$scope.charmOptions = {
  "B": {"stat": "1 str", statVar: "strength", statValue: 1, charmName: "Uruz"},
  "C": {"stat": "1 min" ...etc},
  "D": {"stat": "1 dex" ...etc},
  ...etc
```

Instead of:
```
$scope.charmOptions = [
  {"charm": "B", stat: "1 str", ...etc},
  ...etc
];
```

The difference being that now you can index into `$scope.charmOptions` to find the charm you want. (`$scope.charmOptions["B"]`) This is much faster than looping through the charm options to find the one with the right charm.

Reply:
This is working with the new charmOptions layout now. I also added an ID to use in the ng-repeat function, "option.id". I'm sure theres a better way of doing that.